// express-route-by-jsdoc/index.js
/**
 * generate express router by jsDoc
 * @param {object} app express Object
 * @param {object} option option
 */
function routeGenerator (app = require('express')(), options = {}) {
  const expressSwagger = require('express-swagger-generator')(app)
  const routerCreator = require('./routecreator')
  if (options.docs) {
    options.route = options.docs
    console.log('Swagger docs : %s://%s%s', options.swaggerDefinition.schemes[0], options.swaggerDefinition.host, options.docs.url)
    expressSwagger(options)
  }
  // const basePath = options.swaggerDefinition ? options.swaggerDefinition.basePath : '/'
  // const jsPath = options.api ? options.api.jspath : './src/'

  const postProcess = options.api ? options.api.postprocess : (req, res, next) => {
    var patten =
    {
      response: {
        result: 'true',
        remark: 'success',
        code: res.httpcode || 200
      },
      data: res.body
    }
    res.status(res.httpcode || 200).json(patten)
    next()
  }
  // const r =
  routerCreator(app, postProcess, options)
}

module.exports = routeGenerator
